sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -102,10 +102,9 @@ public void registerUdaf(String functionName, Combine.CombineFn combineFn) {
     defaultSchema.add(functionName, new UdafImpl(combineFn));
   }
 
-  public PTransform<PCollectionTuple, PCollection<Row>> parseQuery(String query)
-      throws ParseException {
+  public PCollection<Row> parseQuery(Pipeline pipeline, String query) throws ParseException {
     try {
-      return planner.convertToBeamRel(query).toPTransform();
+      return BeamSqlRelUtils.toPCollection(pipeline, planner.convertToBeamRel(query));
```

Comment:
Why does parser need to know about the pipeline? Can this logic happen outside of `BeamSqlEnv`? I would rather have another abstraction returned from `parseQuery` instead of passing a mutable global state into some blackbox. For example, I am thinking we can return something like `SqlTransform` here, if we could construct it with an instance of current `BeamSqlEnv`

Reply:
This is not the parser, this is the function to convert from a BeamRelNode to a Beam Pipeline. I think you are intending to comment 3 or 4 lines up?

Reply:
Yes, I meant BeamSqlEnv.parseQuery(). My GitHub highlighting skills are bad

Reply:
It makes the API a bit clunkier (all uses call both functions back to back) but I refactored to split the two operations.